File: baekjoon/행렬.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	scanner = bufio.NewScanner(os.Stdin)
	out     = bufio.NewWriter(os.Stdout)
)

var (
	height, width int
	original      [][]int
	destination   [][]int
)

func check(orig [][]int, dest [][]int) bool {
	for i := range orig {
		for j := range orig[0] {
			if orig[i][j] != dest[i][j] {
				return false
			}
		}
	}
	return true
}

func flip(matrix [][]int, h, w int) {
	for i := h; i < h+3; i++ {
		for j := w; j < w+3; j++ {
			if matrix[i][j] == 1 {
				matrix[i][j] = 0
			} else {
				matrix[i][j] = 1
			}
		}
	}
}

func main() {
	defer out.Flush()

	scanner.Scan()
	fmt.Sscanf(scanner.Text(), "%d %d\n", &height, &width)

	original = make([][]int, height)
	destination = make([][]int, height)

	for i := 0; i < height; i++ {
		scanner.Scan()
		split := strings.Split(scanner.Text(), "")
		original[i] = make([]int, width)
		for j := 0; j < width; j++ {
			original[i][j], _ = strconv.Atoi(split[j])
		}
	}

	for i := 0; i < height; i++ {
		scanner.Scan()
		split := strings.Split(scanner.Text(), "")
		destination[i] = make([]int, width)
		for j := 0; j < width; j++ {
			destination[i][j], _ = strconv.Atoi(split[j])
		}
	}

	if check(original, destination) {
		fmt.Fprintln(out, 0)
		return
	}

	if height < 3 || width < 3 {
		fmt.Fprintln(out, -1)
		return
	}

	cnt := 0

	for i := 0; i < height-2; i++ {
		for j := 0; j < width-2; j++ {
			if destination[i][j] != original[i][j] {
				flip(original, i, j)
				cnt++
			}
		}
	}

	if !check(original, destination) {
		fmt.Fprintln(out, -1)
	} else {
		fmt.Fprint(out, cnt)
	}

}

File: baekjoon/행렬.js
const [size, ...m] = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const flip = (matrix, h, w) => {
  for (let i = h; i < h + 3; i++) {
    for (let j = w; j < w + 3; j++) {
      matrix[i][j] = matrix[i][j] ? 0 : 1;
    }
  }
};

const main = () => {
  const [h, w] = size.split(" ").map(Number);

  if (m.slice(0, h).join("") === m.slice(h).join("")) return 0;

  if (h < 3 || w < 3) return -1;

  const orig = m.slice(0, h).map((v) => v.split("").map(Number));
  const dest = m.slice(h).map((v) => v.split("").map(Number));

  let count = 0;
  for (let i = 0; i < h - 2; i++) {
    for (let j = 0; j < w - 2; j++) {
      if (orig[i][j] !== dest[i][j]) {
        flip(orig, i, j);
        count++;
      }
      if (
        j == w - 3 &&
        (orig[i][j + 1] !== dest[i][j + 1] || orig[i][j + 2] !== dest[i][j + 2])
      )
        return -1;
      if (
        i == h - 3 &&
        (orig[i + 1][j] !== dest[i + 1][j] || orig[i + 2][j] !== dest[i + 2][j])
      )
        return -1;
      if (
        i == h - 3 &&
        j == w - 3 &&
        (orig[i + 1][j + 1] !== dest[i + 1][j + 1] ||
          orig[i + 2][j + 2] !== dest[i + 2][j + 2])
      )
        return -1;
    }
  }
  return count;
};

console.log(main());
